tests/sources/test_redis.py

```diff
@@ -0,0 +1,199 @@
+        return [1, 2, 3]
+
+    async def config_get(self, databases):
+        return {"databases": "1"}
+
+    async def ping(self):
+        return False
+
+    async def aclose(self):
+        return True
+
+    async def type(self, key):  # NOQA
+        return "string"
+
+    async def memory_usage(self, key):
+        return 10
+
+    async def validate_database(self, db=0):
+        await self.execute_command()
+        await self.ping()
+
+
+@asynccontextmanager
+async def create_redis_source():
+    async with create_source(
+        RedisDataSource,
+        host="localhost",
+        port=6379,
+        database="0",
+        username="username",
+        password="password",
+    ) as source:
+        yield source
+
+
+@pytest.mark.asyncio
+async def test_ping_positive():
+    async with create_redis_source() as source:
+        source.client.ping = AsyncMock()
+        await source.ping()
+
+
+@pytest.mark.asyncio
+async def test_ping_negative():
+    async with create_redis_source() as source:
+        with pytest.raises(Exception):
+            await source.ping()
+
+
+@pytest.mark.asyncio
+async def test_validate_config_when_database_type():
+    async with create_redis_source() as source:
+        source.client.database = ["1", "db123", "123"]
+        source.ping = AsyncMock(return_value=True)
+        with pytest.raises(ConfigurableFieldValueError):
+            await source.validate_config()
+
+
+@pytest.mark.asyncio
+async def test_validate_config_when_database_is_invalid():
+    async with create_redis_source() as source:
+        source.client.database = ["123"]
+        with pytest.raises(ConfigurableFieldValueError):
+            await source.validate_config()
+
+
+@pytest.mark.asyncio
+@freeze_time("2023-01-24T04:07:19+00:00")
+async def test_get_docs():
+    async with create_redis_source() as source:
+        source.client.database = [1]
+
+        with mock.patch(
+            "redis.from_url",
+            return_value=AsyncMock(),
+        ):
+            source.get_db_records = AsyncIterator(items=DOCUMENT)
+            async for (doc, _) in source.get_docs():
+                assert doc in DOCUMENT
+
+
+@pytest.mark.asyncio
+async def test_get_databases_for_multiple_db():
+    async with create_redis_source() as source:
+        source.client.database = [1, 2]
+        async for database in source.client.get_databases():
+            assert database in [1, 2]
+
+
+@pytest.mark.asyncio
+async def test_get_databases_with_astric():
+    async with create_redis_source() as source:
+        source.client.database = ["*"]
+        source.client._client = RedisClientMock()
+        async for database in source.client.get_databases():
+            assert database == 0
+
+
+@pytest.mark.asyncio
+async def test_get_databases_negative():
+    async with create_redis_source() as source:
+        source.client.database = ["*"]
+        async for database in source.client.get_databases():
+            assert database == []
```

Comment:
It seems this passed because it can't get any result from non-existing `localhost`.